src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +452,317 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+        /// the resulting snippet; otherwise, <see langword="false"/> if the method name and parentheses are assumed to
+        /// already exist and the template will only specify the argument placeholders.</param>
+        /// <param name="parameters">The parameters to the method. If the specific target of the invocation is not
+        /// known, an empty array may be passed to create a template with a placeholder where arguments will eventually
+        /// go.</param>
+        /// <returns></returns>
+        private static XDocument CreateSnippet(string methodName, bool includeMethod, ImmutableArray<IParameterSymbol> parameters, CancellationToken cancellationToken)
+        {
+            XNamespace snippetNamespace = "http://schemas.microsoft.com/VisualStudio/2005/CodeSnippet";
+
+            var template = new StringBuilder();
+
+            if (includeMethod)
+            {
+                template.Append(methodName).Append('(');
+            }
+
+            var declarations = new List<XElement>();
+            foreach (var parameter in parameters)
+            {
+                if (declarations.Any())
+                {
+                    template.Append(", ");
+                }
+
+                // Create a snippet field for the argument. The name of the field matches the parameter name, and the
+                // default value for the field is provided by a call to the internal ArgumentValue snippet function. The
+                // parameter to the snippet function is a serialized SymbolKey which can be mapped back to the
+                // IParameterSymbol.
+                template.Append('$').Append(parameter.Name).Append('$');
+                declarations.Add(new XElement(
+                    snippetNamespace + "Literal",
+                    new XAttribute(snippetNamespace + "Editable", "true"),
+                    new XElement(snippetNamespace + "ID", new XText(parameter.Name)),
+                    new XElement(snippetNamespace + "Function", new XText($"ArgumentValue({SymbolKey.CreateString(parameter, cancellationToken)})"))));
+            }
+
+            if (!declarations.Any())
+            {
+                // If the invocation does not have any parameters, include an empty placeholder in the snippet template
+                // to ensure the caret starts inside the parentheses and can track changes to other overloads (which may
+                // have parameters).
+                template.Append("$placeholder$");
+                declarations.Add(new XElement(
+                    snippetNamespace + "Literal",
+                    new XAttribute(snippetNamespace + "Editable", "true"),
+                    new XElement(snippetNamespace + "ID", new XText("placeholder")),
+                    new XElement(snippetNamespace + "ToolTip", new XText("")),
+                    new XElement(snippetNamespace + "Default", new XText("")),
+                    new XElement(snippetNamespace + "Function")));
```

Comment:
interesting... what does it mean to have a `Function` that doesn't actually point to a funtion?  Is this necessary to get this to work?  Or could this element be removed?

Reply:
If `Function` is omitted or empty, it uses `Default`.

Reply:
⏱ will remove this use of `Function`